File: scripts/verify_pricing.py
"""
Cross-checks the Rust port (crates/nse_pricing) against the real
box_pricer.py / interest_calc.py by running both over the same test
vectors in fixtures/pricing_test_vectors.json, printing one JSON line per
vector in a shape meant to be diffed against:

    cargo run -p nse_pricing --example verify_pricing

Run from anywhere; paths are resolved relative to this file, not the cwd.

Usage:
    python scripts/verify_pricing.py
"""

import json
import os
import sys
from datetime import date, timedelta

REPO_ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..")
sys.path.insert(0, REPO_ROOT)

from box_pricer import calculate_box_sell_price, calculate_box_long_price  # noqa: E402
from interest_calc import calculate_interest_rate  # noqa: E402

FIXTURE = os.path.join(REPO_ROOT, "fixtures", "pricing_test_vectors.json")


def main():
    with open(FIXTURE, encoding="utf-8") as f:
        vectors = json.load(f)

    for v in vectors:
        k1, k2 = v["k1"], v["k2"]
        days = v["days_to_expiry"]

        # calculate_interest_rate reads date.today() internally, so build an
        # expiry string that lands exactly `days` days out *right now* --
        # keeps this deterministic without touching interest_calc.py.
        expiry = (date.today() + timedelta(days=days)).strftime("%d%b%Y").upper()

        quotes = {
            (k1, "CE"): {"bid": v["call_k1"]["bid"], "ask": v["call_k1"]["ask"], "display_name": f"CE{k1}"},
            (k1, "PE"): {"bid": v["put_k1"]["bid"], "ask": v["put_k1"]["ask"], "display_name": f"PE{k1}"},
            (k2, "CE"): {"bid": v["call_k2"]["bid"], "ask": v["call_k2"]["ask"], "display_name": f"CE{k2}"},
            (k2, "PE"): {"bid": v["put_k2"]["bid"], "ask": v["put_k2"]["ask"], "display_name": f"PE{k2}"},
        }

        row = {"name": v["name"], "k1": k1, "k2": k2}

        try:
            short_spread, _ = calculate_box_sell_price(quotes, k1, k2)
            row["short_spread"] = short_spread
            row["short_interest_rate"] = calculate_interest_rate(short_spread, k1, k2, expiry)["interest_calculation"]
        except ValueError:
            row["short_spread"] = None
            row["short_interest_rate"] = None

        try:
            long_spread, _ = calculate_box_long_price(quotes, k1, k2)
            row["long_spread"] = long_spread
            row["long_interest_rate"] = calculate_interest_rate(long_spread, k1, k2, expiry)["interest_calculation"]
        except ValueError:
            row["long_spread"] = None
            row["long_interest_rate"] = None

        print(json.dumps(row))


if __name__ == "__main__":
    main()

File: crates/nse_pricing/examples/verify_pricing.rs
//! Cross-check tool: runs the same test vectors from
//! `fixtures/pricing_test_vectors.json` through the real Rust formulas
//! (`box_pricer` + `interest_calc`) and prints one JSON line per vector, in
//! the same shape `scripts/verify_pricing.py` prints its real-Python
//! results in -- run both and diff:
//!
//!     cargo run -p nse_pricing --example verify_pricing
//!     python scripts/verify_pricing.py
//!
//! Any mismatch means the Rust port has drifted from `box_pricer.py`/
//! `interest_calc.py`, not just that a hand-picked unit test happens to
//! agree with itself.
//!
//! Also prints `nse_pricing::cost`'s real, tax-adjusted box price (using
//! `CostConfig::default()`'s buy/sell round-trip multipliers) alongside the
//! raw pre-tax spread above. There's no Python side to diff this against.

use nse_adapt::QuoteUpdate;
use nse_pricing::{CostConfig, FourLegs, box_pricer, cost, interest_calc};
use serde::Deserialize;
use serde_json::json;

#[derive(Deserialize)]
struct Leg {
    bid: f64,
    ask: f64,
}

#[derive(Deserialize)]
struct Vector {
    name: String,
    k1: f64,
    k2: f64,
    days_to_expiry: i64,
    call_k1: Leg,
    put_k1: Leg,
    call_k2: Leg,
    put_k2: Leg,
}

fn leg_quote(leg: &Leg) -> QuoteUpdate {
    let mut q = QuoteUpdate::from_enhanced_mbp(0, &[], 0);
    q.bid_price[0] = (leg.bid * 100.0).round() as i64;
    q.ask_price[0] = (leg.ask * 100.0).round() as i64;
    q
}

fn main() {
    let path = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../fixtures/pricing_test_vectors.json"
    );
    let text =
        std::fs::read_to_string(path).unwrap_or_else(|e| panic!("failed to read {path}: {e}"));
    let vectors: Vec<Vector> =
        serde_json::from_str(&text).expect("invalid pricing_test_vectors.json");
    let cost_cfg = CostConfig::default();

    for v in vectors {
        let strike_difference = v.k2 - v.k1;

        let short_spread =
            box_pricer::box_sell_price(v.call_k1.bid, v.put_k1.ask, v.call_k2.ask, v.put_k2.bid);
        let long_spread =
            box_pricer::box_long_price(v.call_k1.ask, v.put_k1.bid, v.call_k2.bid, v.put_k2.ask);

        let short_interest_rate = short_spread.and_then(|p| {
            interest_calc::annualized_interest_rate(p, strike_difference, v.days_to_expiry)
        });
        let long_interest_rate = long_spread.and_then(|p| {
            interest_calc::annualized_interest_rate(p, strike_difference, v.days_to_expiry)
        });

        let (call_k1, put_k1, call_k2, put_k2) = (
            leg_quote(&v.call_k1),
            leg_quote(&v.put_k1),
            leg_quote(&v.call_k2),
            leg_quote(&v.put_k2),
        );
        let legs = FourLegs {
            call_k1: Some(&call_k1),
            call_k2: Some(&call_k2),
            put_k1: Some(&put_k1),
            put_k2: Some(&put_k2),
        };
        let long_spread_tax = cost::long_box_taxed_price(&legs, &cost_cfg);
        let short_spread_tax = cost::short_box_taxed_price(&legs, &cost_cfg);

        let row = json!({
            "name": v.name,
            "k1": v.k1,
            "k2": v.k2,
            "short_spread": short_spread,
            "short_interest_rate": short_interest_rate,
            "short_spread_tax": short_spread_tax,
            "long_spread": long_spread,
            "long_interest_rate": long_interest_rate,
            "long_spread_tax": long_spread_tax,
        });
        println!("{row}");
    }
}
